### PR TITLE
Plugins: Automatically forward plugin request HTTP headers in outgoing HTTP requests

### LINKED ISSUE #57065: Fix or document handling of non-datasource-specific-headers in backend-datasources
the alerting module currently sets the `FromAlert` and `X-Cache-Skip` values in `backend.QueryDataRequest.Headers` ,and, it seems, datasource-plugin are supposed to forward this using http-headers to the databases.

for a developer of a datasource-plugin it's not clear how to handle the situation:
- should they look for these specific values in `backend.QueryDataRequest.Headers`  and send them "out"? this seems brittle. if this is the way to go, we should document these special header values, and write some sample code to demonstrate the best practice
- should this happen somehow automatically by generic grafana code?

### PR DESCRIPTION
**Pull request by marefr**
**What is this feature?**
Automatically forward core plugin request HTTP headers in outgoing HTTP requests.

Details:
- Introduces constants for the `FromAlert` alerting header and `X-Cache-Skip` header.
- FromAlert plugin request header is forwarded in outgoing HTTP requests in its canonical form, `Fromalert`.

**Why do we need this feature?**
Core datasource plugin authors don't have to specifically handle forwarding of HTTP headers, e.g. do not have to "hardcode" the header-names in the datasource plugin, if not having custom needs.

**Who is this feature for?**
Core datasource plugin authors

**Which issue(s) does this PR fix?**:
Fixes #57065 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by marefr** [comment]
@alexweav adding you as reviewer since you made the change to enable Loki forwarding of `Fromalert` HTTP header in outgoing HTTP request, ref https://github.com/grafana/grafana/pull/56896

**Review by gabor — COMMENTED** [review]
the Loki changes are ok 👍

**Comment by gabor** [comment]
@marefr i did some testing for Loki+CallResource, and it seems the `Cookie` header that's sent from the browser to the grafana-server  (so basically the grafana-session-id) , is sent to the Loki database. shouldn't we remove the `Cookie` header there?

**Comment by marefr** [comment]
> @marefr i did some testing for Loki+CallResource, and it seems the `Cookie` header that's sent from the browser to the grafana-server (so basically the grafana-session-id) , is sent to the Loki database. shouldn't we remove the `Cookie` header there?

@gabor thanks. Interesting. Cookies should only be forwarded if configured in the datasource settings, but the  grafana-session-id should never be forwarded. I'll look into it

**Comment by svennergr** [comment]
> the Loki changes are ok 👍

Just to be completely sure, I checked if compressing (`Accept-Encoding`) and forward OAuth identity (including `X-ID-Token`) is working in Loki. LGTM - great stuff @marefr!

**Comment by marefr** [comment]
@gabor pushed a fix for forwarding of cookies. Really good find, thanks. Also found that User-Agent is not properly set and is now fixed as well. 

Planning to add some more validation to the backendplugin integration tests just to be sure. One odd thing I found for Prometheus query data is that it has a header `Accept-Encoding: gzip` set, but I can't understand right now where or how that is set :confused:

**Comment by svennergr** [comment]
> One odd thing I found for Prometheus query data is that it has a header `Accept-Encoding: gzip` set, but I can't understand right now where or how that is set 😕

Might very well be the auto compression of Go's http transport. It's enabled as long as you don't set `DisableCompression` in https://pkg.go.dev/net/http#Transport.

**Review by xnyo — CHANGES_REQUESTED** [review]
Awesome! :clap::clap::clap::clap:
Left some comments for a docstring change:

**Comment by kostrse** [comment]
No regressions with Azure auth

File: pkg/services/pluginsintegration/clientmiddleware/user_header_middleware.go [hunk 1]
```diff
@@ -27,41 +24,25 @@ type UserHeaderMiddleware struct {
...
-		httpHeaders := http.Header{
-			proxyutil.UserHeaderName: []string{reqCtx.Login},
-		}
-
-		middlewares = append(middlewares, httpclientprovider.SetHeadersMiddleware(httpHeaders))
-	} else {
-		middlewares = append(middlewares, httpclientprovider.DeleteHeadersMiddleware(proxyutil.UserHeaderName))
-	}
-
-	ctx = sdkhttpclient.WithContextualMiddleware(ctx, middlewares...)
```
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/services/pluginsintegration/clientmiddleware/user_header_middleware.go:54]: Fixed

**Review by xnyo — APPROVED** [review]
LGTM! :rocket:

**Review by itsmylife — APPROVED** [review]
🎉

**Pull request merged**